### PR TITLE
Use QueryIndex() for better performance in OC.PublishLater

### PR DESCRIPTION
**Pull request by hishamco**
Address the comment [here](https://github.com/OrchardCMS/OrchardCore/pull/10935#issuecomment-1221855101)

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Indexes/PublishLaterPartIndexProvider.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Indexes/PublishLaterPartIndexProvider.cs:75]: Published then Latest, both at the end.

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:28]: You should update the Create method to include new column, so that a fresh setup defines all what is needed in one pass. And with Published then Latest, both at the end.

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:34]: Idem and with the new included column `Id`.

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:38]: Should now return 3 so that a fresh setup only execute in one pass the updated Create method.

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Services/ScheduledPublishingBackgroundTask.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Services/ScheduledPublishingBackgroundTask.cs:43]: So here we may want to set the schedule date to null, and then remove the related handler, this if we use the same logic as in ArchiveLater.

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:66]: Here you should not create an index but only add the columns that were missings, e.g. Published and Latest at the end, something like this for each column.

        SchemaBuilder.AlterIndexTable<PublishLaterPartIndex>(table => table
            .AddColumn<bool>("xxx")
        );

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:80]: Here you should first drop the existing non clustured IDX_ index, something like this.

        SchemaBuilder.AlterIndexTable<PublishLaterPartIndex>(table => table
            .DropIndex("IDX_...")
        );

Then rebuild it with all the inculded columns, Published/Latest at the end.

Note: In one project I needed to drop the non clustered IDX_ at the very beginning, but I think it was needed before removing columns (included in this IDX_) of the related Index table, which is not the case here.

Anyway would need to be tried, first with a fresh setup without the changes in this PR, then restart the app to see what happens when this migration step is executed. Then finally re-do a fresh setup with all the updates to see if the new Create method is okay. At least with Sqlite and Sql Server.

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:66]: You are right, I was looking for that at the time, seems I typed a wrong key stroke ;)

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:80]: Again you are right, it's my fingers fault ;)

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:80]: No problem ;)

Will review again when you will update it, but take your time, no pressure.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:80]: I will update them now, I already switch to branch but waiting for VS to load the projects ;)

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:28]: I didn't touch the Create method for backward compatibility

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:66]: Seems I was looking for `AlterMapIndexTable` ;)

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:34]: Why?

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:80]: @jtkech if we drop the index, then we can create the columns from ground up, am I right instead of adding columns using `AddColumn()`

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:28]: This is wrong, on an existing site/database the `Create()` method is not executed again.

Then on a fresh setup better to only execute an updated `Create()` method that returns the last version number, this in place of executing the old one and then migration steps that may add/remove things, particularly when migration steps may fail with some DB providers.

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:34]: As we did for ArchiveLater and that I commented   https://github.com/OrchardCMS/OrchardCore/pull/10935#issuecomment-1221855101

See also https://github.com/OrchardCMS/OrchardCore/pull/10935#issuecomment-1236189580 for more details about non clustered indexes.

So here because we want to use `QueryIndex()` that returns all columns of the index Table including the `Id` column, good to also include the `Id` column in this non clustered index (which is an inner index of the index Table), so that we have a better chance to not hit the DB clusters.

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:66]: See my other comments for more details

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:80]: Yes, you have to use `.AddColumn()`, then also `.DropIndex()` to first remove the inner non clustered index, see my other comments for more details, e.g. https://github.com/OrchardCMS/OrchardCore/pull/10935#issuecomment-1221855101, then finally `.CreateIndex("IDX_Something)` to rebuild the inner index by including all columns.

**Comment by jtkech** [comment]
@hishamco 

I answered to your comments but sorry I submitted them as a review, the "problem" is that here we only see my answers, not your related comments that you may have to retrieve ;)

**Comment by hishamco** [comment]
> This is wrong, on an existing site/database the Create() method is not executed again.

Seems my understaning for `Create` is wrong ;)

**Comment by sebastienros** [comment]
This one is waiting for merge conflict fixes and an approval

**Comment by sebastienros** [comment]
Why is it blocking 1.5 release?

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Services/ScheduledPublishingBackgroundTask.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Services/ScheduledPublishingBackgroundTask.cs:46]: Check if part is null as we do for archiveLater

            var part = contentItem.As<ArchiveLaterPart>();
            if (part != null)
            {
                part.ScheduledArchiveUtc = null;
            }

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:83]: @hishamco I will approve but first did you try this migration step.

I mean do a setup before the code changes made in this PR, then check that the feature works, then update the code as done in this PR, ensure  that this migration step is run and works, and then that the feature still works.

**Comment by jtkech** [comment]
@hishamco @sebastienros 

Just fixed the conflicts and applied the suggested part null check as in archiveLater.

I will try soon if the migration step works well, than I will approve.

**Comment by jtkech** [comment]
@hishamco Just tried with Sqlite.

Look at my last commits for more details on the following fixes.

- You need separate `SchemaBuilder.AlterIndexTable()` for dropping and then re-creating the IDX_ index (the inner sql index of the IndexTable), then if well dropped you can use the same name to re-create it.

- In the bgTask after setting to null `part.ScheduledPublishUtc` we need to call  `part.Apply()`. I applied the same update for the archiveLater bgTask.

- Also because here we check for non published and latest items, when using the contentManager we need to call `.GetAsync(item.ContentItemId, VersionOptions.Latest)` to be published, by default it returns the published version, which is okay for archiveLater.

- I will try it soon on Sql server.

Also @Piedone in the item editor for publishLater buttons here we deal with the `submit.Save` action, which is good as we save a draft (latest and non published) that will be published by the bgTask.

But just saw that we do the same for archiveLater buttons, so after clicking on the button, you need to re-edit the item to publish it, so that it is retrieved as a latest an publised item to archive.

There are solutions but first is it intended?

**Comment by hishamco** [comment]
Thanks @jtkech seems I replied after you test it in SQLite, if you need me to test anything please let me know

**Comment by jtkech** [comment]
No that's okay, only waiting for @Piedone for the above question in my previous comment about archiveLater that we could tweak here.

**Comment by Piedone** [comment]
The Archive Later behavior doesn't seem to be deliberate, no. But Hisham, you've written that, please chime in.

**Comment by jtkech** [comment]
@hishamco Yes `OC.ArchiveLater` is very similar but is doing the opposite ;)

It filters published items to be unpublished, in place of not published items to be published.

In the same way the buttons should deal with `submit.Publish` in place of `submit.Save`. I asked to @Piedone just to be sure it was not intended.

> @jtkech can I move the OC.ArchiveLater related changes into another PR so I can work on them and let see what's remaining on this PR

No that's okay, for now I only added the missing `part.Apply()` as here, and I just need to replace 2 `submit.Save` by `submit.Publish`. Already done and tested locally, will push it and then approve.

File: src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: If we do so, we need to change this too

https://github.com/OrchardCMS/OrchardCore/blob/242115849b741325b667050b40b78d947ea477f0/src/OrchardCore.Modules/OrchardCore.PublishLater/Drivers/PublishLaterPartDisplayDriver.cs#L57

Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: If we do so, we need to change this too

https://github.com/OrchardCMS/OrchardCore/blob/242115849b741325b667050b40b78d947ea477f0/src/OrchardCore.Modules/OrchardCore.PublishLater/Drivers/PublishLaterPartDisplayDriver.cs#L57

Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: If we do so, we need to change this too

https://github.com/OrchardCMS/OrchardCore/blob/242115849b741325b667050b40b78d947ea477f0/src/OrchardCore.Modules/OrchardCore.PublishLater/Drivers/PublishLaterPartDisplayDriver.cs#L57

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: No, the change only concern archiveLater

Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: If we do so, we need to change this too

https://github.com/OrchardCMS/OrchardCore/blob/242115849b741325b667050b40b78d947ea477f0/src/OrchardCore.Modules/OrchardCore.PublishLater/Drivers/PublishLaterPartDisplayDriver.cs#L57

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: No, the change only concern archiveLater

Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: If we do so, we need to change this too

https://github.com/OrchardCMS/OrchardCore/blob/242115849b741325b667050b40b78d947ea477f0/src/OrchardCore.Modules/OrchardCore.PublishLater/Drivers/PublishLaterPartDisplayDriver.cs#L57

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: No, the change only concern archiveLater

**Review by jtkech — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: If we do so, we need to change this too

https://github.com/OrchardCMS/OrchardCore/blob/242115849b741325b667050b40b78d947ea477f0/src/OrchardCore.Modules/OrchardCore.PublishLater/Drivers/PublishLaterPartDisplayDriver.cs#L57

Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: If we do so, we need to change this too

https://github.com/OrchardCMS/OrchardCore/blob/242115849b741325b667050b40b78d947ea477f0/src/OrchardCore.Modules/OrchardCore.PublishLater/Drivers/PublishLaterPartDisplayDriver.cs#L57

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: BTW this will break thing, because we should change the name in view

https://github.com/OrchardCMS/OrchardCore/blob/242115849b741325b667050b40b78d947ea477f0/src/OrchardCore.Modules/OrchardCore.ArchiveLater/Views/ArchiveLaterPart.Edit.cshtml#L18

I can fix it in both `OC.ArchiveLater` and `OC.PublishLater`

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: Why? we can you `Publish` here, 'Unpublish` there

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: No

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: Could you please elaborate what's going on? How you change the submit button name in the driver without changing the view, I'm confused ;)

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.ArchiveLater/Drivers/ArchiveLaterPartDisplayDriver.cs:52]: Normally I changed it in the view too, I think you commented before I submitted my changes

**Pull request merged**
